service/storage_service.cc

```diff
@@ -2886,9 +2887,8 @@ future<> storage_service::update_peer_info(gms::inet_address endpoint) {
     if (!ep_state) {
         co_return;
     }
-    for (auto& entry : ep_state->get_application_state_map()) {
-        auto& app_state = entry.first;
-        auto& value = entry.second;
+    auto state_map = ep_state->get_application_state_map();
+    for (auto& [app_state, value] : state_map) {
```

Comment:
Gossiper is supposed to serialize execution of listeners via the lock_endpoint() thing.

It's not safe to allow listeners to run concurrently, since they may get reordered. For example, the call which updates system tablets below could overwrite a more recent update.

Reply:
Who is modifying endpoint states concurrently?

Reply:
> Gossiper is supposed to serialize execution of listeners via the lock_endpoint() thing.

Where is it documented?

> It's not safe to allow listeners to run concurrently, since they may get reordered. For example, the call which updates system tablets below could overwrite a more recent update.

The problem with relying on lock_endpoint is possible deadlocks if the notified function calls back into gossiper, e.g. force_remove_endpoint that calls evict_from_membership that acquires the endpoint lock as well.

See https://github.com/scylladb/scylladb/pull/14471#issuecomment-1620102615
But I've hit a similar deadlock in my branch from with the `topology.test_cluster_features` as well when adding lock_endpoint in real_mark_alive to serialize changes to `_live_endpoints`, `_unreachable_endpoints` + `replicate_live_endpoints_on_change` that yields, requiring the lock. Plus, it calls the `on_alive` notifications that, as you said are supposed to be called under that lock.

> Who is modifying endpoint states concurrently?

I hit it in my branch that includes https://github.com/scylladb/scylladb/commit/33da21c826e44d48e8234e3c9368aa7a75a8f2d7
I'm not sure what exactly who was modifying the endpoint states concurrently.
The current design is quite fragile and under-documented.

If we always hold the endpoint lock when calling notifications we may need to expose entry points that can be called safely when the lock is held and avoid acquiring it in the gossiper layer.